File: Python/0017-letter-combinations-of-a-phone-number.py
# Time:  O(4^n)
# Space: O(n)
# Recursive Solution
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        self.lookup = { '2':['a', 'b', 'c'], '3':['d', 'e', 'f'], '4':['g', 'h', 'i'], '5':['j', 'k', 'l'] , '6':['m', 'n', 'o'], '7':['p', 'q', 'r', 's'], '8':['t', 'u', 'v'], '9':['w', 'x', 'y', 'z']}
        self.result = []
        self.helper(0, "", digits)
        return self.result

    def helper(self, cur, ans, digits):
        if cur == len(digits):
            self.result.append(ans)
            return

        for c in self.lookup[digits[cur]]:
            self.helper(cur+1, ans + c, digits)


# Time:  O(4^n)
# Space: O(4^n)
# BFS Solution using two lists
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        mapping = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl",\
                  "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}
        currlevel = ['']
        for d in digits:
            nextlevel = []
            for x in currlevel:
                for c in mapping[d]:
                    nextlevel.append(x + c)
            currlevel = nextlevel
        return currlevel


# Time:  O(4^n)
# Space: O(4^n)
# BFS Solution using queue
import collections
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        mapping = {"2": "abc", "3": "def", "4": "ghi", "5": "jkl",\
                  "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"}
        queue = collections.deque([""])
        for i, d in enumerate(digits):
            # i represents no of chars in prev level
            while len(queue[0]) == i:
                x = queue.popleft()
                for c in mapping[d]:
                    queue.append(x + c)
        return queue

File: C++/0017-letter-combinations-of-a-phone-number.cpp
// Time:  O(4^n)
// Space: O(4^n)
// BFS
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty())
            return {};
        unordered_map<char, string> mp {{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"},{'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
        // queue<string> q;
        // q.push("");
        // for (size_t i = 0; i < digits.length(); ++i) {
        //     while (q.front().length() == i) {
        //         string path = q.front(); q.pop();
        //         for (char c: mp[digits[i]]) {
        //             q.push(path + c);
        //         }
        //     }
        // }
        vector<string> cur {""}, next;
        for (size_t i = 0; i < digits.length(); ++i) {
            for (auto x: cur)
                for (char c: mp[digits[i]])
                    next.push_back(x + c);
            cur = next;
            next.clear();
        
        }
        return cur;
    }
};
